## Ebay Car Sales Exploration



We will be working on a dataset of used cars from *eBay Kleinanzeigen*, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data).  The version of the dataset we are working with is a sample of 50,000 data points that was prepared by [Dataquest](https://www.dataquest.io) including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which which year the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which which year the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.


The aim of this project is to clean the data and analyze the included used car listings.

### 1.1 Exploring German Ebay classifieds dataset.

### Dataset scraped from Kaggle and dirtied to reflect real world datasets.

### Sampled 50,000 data points from the original dataset.

In [2]:
import pandas as pd
import numpy as np


In [3]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [4]:
autos['offerType'].value_counts()

Angebot    49999
Gesuch         1
Name: offerType, dtype: int64

In [5]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Changes being made are:
- Change column names from camelcase to snakecase
- Renaming a few columns to accurately describe the columns

In [7]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test', 
                     'vehicle_type', 'registeration_year', 'gearbox', 'power_PS', 'model', 
                     'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 
                     'ad_created','picture_count', 'postal_code', 'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registeration_year  \
0  privat    Angebot  $5,000  control          bus                2004   
1  privat    Angebot  $8,500  control    limousine                1997   
2  privat    Angebot  $8,990     test    limousine                2009   
3  privat    Angebot  $4,350  control   kleinwagen                2007   
4  privat    Angebot  $1,350     test        kombi                2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  picture_count  \
0     peugeot              nein  2016-03-26 00:00:00              0   
1         bmw              nein  2016-04-04 00:00:00              0   
2  volkswagen              nein  2016-03-26 00:00:00              0   
3       smart              nein  2016-03-12 00:00:00              0   
4        ford              nein  2016-04-01 00:00:00              0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### 1.2 Basic Exploration

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registeration_year  gearbox      power_PS  model  \
count         44905        50000.000000    47320  50000.000000  47242   
unique            8                 NaN        2           NaN    245   
top       limousine                 NaN  manuell           NaN   golf   
freq          12859                 NaN    36993           NaN   4024   
mean            NaN         2005.073280      NaN    116.355920    NaN   
std             NaN          105.712813      NaN    209.216627    NaN   
min             NaN         1000.000000      NaN      0.000000    NaN   
25%             NaN         1999.000000      NaN     70.000000    NaN   
50%             NaN         2003.000000      NaN    105.000000    NaN   
75%             NaN         2008.000000      NaN    150.000000    NaN   
max             NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  picture_count   postal_code            last_seen  
count                 50000        50000.0  50000.000000                50000  
unique                   76            NaN           NaN                39481  
top     2016-04-03 00:00:00            NaN           NaN  2016-04-07 06:17:27  
freq                   1946            NaN           NaN                    8  
mean                    NaN            0.0  50813.627300                  NaN  
std                     NaN            0.0  25779.747957                  NaN  
min                     NaN            0.0   1067.000000                  NaN  
25%                     NaN            0.0  30451.000000                  NaN  
50%                     NaN            0.0  49577.000000                  NaN  
75%                     NaN            0.0  71540.000000                  NaN  
max                     NaN            0.0  99998.000000                  NaN

Initial Observations:

- Many of the columns are text or dates so descriptive statistics are   limited.
- num_photos column looks odd, further investigation is needed.
- There are a number of text columns where all of the values are the 
  same.
    - `seller`
    - `offer_type`

In [9]:
autos["picture_count"].value_counts()

0    50000
Name: picture_count, dtype: int64

From the above observation, `picture_count` column has 0 for every single row, it is safe to drop this column and the other two named `seller` and `offer_type` which mostly have only one value.

## Need to work on section 3

In [10]:
autos = autos.drop(['picture_count', 'seller', 'offer_type'], axis = 1)

Converting `price` and `odometer` columns to numeric by removing extra symbols.

In [11]:
autos['price'] = (autos['price']
                            .str.replace('$', '')
                            .str.replace(',', '')
                            .astype(int)
                 )

In [12]:
autos['odometer'] = (autos['odometer']
                            .str.replace('km', '')
                            .str.replace(',', '')
                            .astype(int)
                 )

Since `odometer` column is measured in Kilo Meters, we are converting them to miles to avoid any confusion.

In [13]:
autos['odometer'] = autos['odometer']*0.6213 
#be careful to run only once
autos.rename({'odometer': 'odometer_miles'}, axis=1, inplace=True)
autos['odometer_miles'].head()

0    93195.0
1    93195.0
2    43491.0
3    43491.0
4    93195.0
Name: odometer_miles, dtype: float64

# Exploring Odometer and Price Columns

In [14]:
 autos['odometer_miles'].value_counts()

93195.0    32424
77662.5     5170
62130.0     2169
55917.0     1757
49704.0     1436
43491.0     1230
37278.0     1164
31065.0     1027
3106.5       967
24852.0      819
18639.0      789
12426.0      784
6213.0       264
Name: odometer_miles, dtype: int64

In [15]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

Looking at the value_counts's output, it could be inferred that the values are rounded to the nearest 100. It is safe to assume that the seller was given a pre-set options to choose from. 

In [28]:
autos[['price', 'registeration_year', 'power_PS', 'odometer_miles',
       'registration_month', 'postal_code']].corr(method='spearman')

price  registeration_year  power_PS  odometer_miles  \
price               1.000000            0.564985  0.551533       -0.374413   
registeration_year  0.564985            1.000000  0.198150       -0.338563   
power_PS            0.551533            0.198150  1.000000        0.001459   
odometer_miles     -0.374413           -0.338563  0.001459        1.000000   
registration_month  0.147102            0.051795  0.130015       -0.012713   
postal_code         0.091271            0.038787  0.096069       -0.029904   

                    registration_month  postal_code  
price                         0.147102     0.091271  
registeration_year            0.051795     0.038787  
power_PS                      0.130015     0.096069  
odometer_miles               -0.012713    -0.029904  
registration_month            1.000000     0.014669  
postal_code                   0.014669     1.000000

Correlation matrix gives us some interesting insight into which features affect pricing the most.
- `registration_year` alone can explain 56% of car's price.
- `power_PS` also seems to be a stron indicator of the price of a car, explains about 55% of price.
- `odometer_miles` explains about 37% of a car's price, number of miles a car travels has an impact on it's price.
- `registration_month` can explain 14% of the pricing information.
- `postal_code` could explain 9% of pricing data on its own.

Although